File: 高级数据结构与算法/Python/第2章 分治法/2.3 快速幂/matrix_fast_exponentiation.py
N = 2

def matrix_multiply(a, b):
    c = [
        [0, 0],
        [0, 0]
    ]
    for i in range(N):
        for j in range(N):
            for k in range(N):
                c[i][j] += a[i][k] * b[k][j]
    return c

def matrix_fast_exp(n):
    result = [
        [1, 1],
        [1, 0]
    ]
    M = [
        [1, 1],
        [1, 0]
    ]

    while n > 0:
        if n & 1:
            result = matrix_multiply(result, M)
        M = matrix_multiply(M, M)
        n >>= 1
    
    return result[0][0]

def main():
    for i in range(1, 11):
        print(matrix_fast_exp(i-2), end= ' ')
    print()

if __name__ == "__main__":
    main()

File: 高级数据结构与算法/Java/第2章 分治法/2.3 快速幂/MatrixFastExponentiation.java
public class MatrixFastExponentiation {
    public static final int N = 2;

    public static void main(String[] args) {
        for(int i = 1; i <= 10; i++) {
            System.out.print(matrixFastExp(i-2) + " ");
        }
        System.out.println();
    }

    public static int[][] matrixMultiply(int[][] a, int[][] b) {
        int[][] c = new int[N][N];
        for(int i = 0; i < N; i++) {
            for(int j = 0; j < N; j++) {
                for(int k = 0; k < N; k++) {
                    c[i][j] += a[i][k] * b[k][j];
                }
            }
        }
        return c;
    }

    public static int matrixFastExp(int n) {
        int[][] result = {
            {1, 1},
            {1, 0}
        };
        int[][] M = {
            {1, 1},
            {1, 0}
        };

        while(n > 0) {
            if((n & 1) != 0) {
                result = matrixMultiply(result, M);
            }
            M = matrixMultiply(M, M);
            n >>= 1;
        }

        return result[0][0];
    }
}
